src/main/scala/common/sodor_internal_tile.scala

```diff
@@ -0,0 +1,145 @@
+package sodor.common
+
+import chisel3._
+import chisel3.util._
+
+import freechips.rocketchip.config._
+import freechips.rocketchip.subsystem._
+import freechips.rocketchip.devices.tilelink._
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.diplomaticobjectmodel.logicaltree.{LogicalTreeNode}
+import freechips.rocketchip.rocket._
+import freechips.rocketchip.subsystem.{RocketCrossingParams}
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.interrupts._
+import freechips.rocketchip.util._
+import freechips.rocketchip.tile._
+import freechips.rocketchip.amba.axi4._
+
+// Event sets for core CSR
+// This object simply disable all performance counters
+object CSREvents {
+  val events = new EventSets(Seq(new EventSet((mask, hit) => false.B, Seq(("placeholder", () => false.B)))))
+}
+
+// Abstract core and tile base class for all cores
+abstract class AbstractCore extends Module {
+  val mem_ports: Seq[MemPortIo]
+  val interrupt: CoreInterrupts
+  val constants: TileInputConstants
+}
+abstract class AbstractInternalTile(implicit val conf: SodorConfiguration) extends Module {
+  val io = IO(new Bundle {
+    val debug_port = Flipped(new MemPortIo(data_width = conf.xprlen))
+    val master_port = Vec(2, new MemPortIo(data_width = conf.xprlen))
+    val interrupt = Input(new CoreInterrupts()(conf.p))
+    val constants = new TileInputConstants()(conf.p)
+  })
+}
+
+// Cores and internal tiles constructors
+trait SodorCoreFactory {
+  def nMemPorts: Int
+  def instantiate(implicit conf: SodorConfiguration): AbstractCore
+}
+
+trait SodorInternalTileFactory {
+  def instantiate(range: AddressSet)(implicit conf: SodorConfiguration): AbstractInternalTile
+}
+
+// Original sodor tile in this repo.
+// This tile is only for 3-stage core since it has a special structure (SyncMem and one memory port)
+class SodorInternalTileStage3(range: AddressSet)(implicit conf: SodorConfiguration) extends AbstractInternalTile
+{
+  val procConst = {
+    class C extends sodor.stage3.constants.SodorProcConstants
+    new C
+  }
+
+  val core   = Module(new sodor.stage3.Core())
+  core.io := DontCare
+  val memory = Module(new SyncScratchPadMemory(num_core_ports = procConst.NUM_MEMORY_PORTS))
+
+  if (procConst.NUM_MEMORY_PORTS == 1) // Only used in stage3
```

Comment:
One of these clauses can be removed, since this tile is only for Stage3?

Reply:
(Outdated, resolved)